### PR TITLE
feat(rust): remote client write data endpoint

### PR DESCRIPTION
**Pull request by wjones127**
* Implements:
  * Add
  * Update
  * Delete
  * Merge-Insert

### CODE REVIEW

File: rust/lancedb/src/remote/table.rs [hunk 1]
```diff
@@ -41,20 +44,47 @@ impl<S: HttpSend> RemoteTable<S> {
...
+                Some(Ok(buffer))
+            }
+            Some(Err(e)) => Some(Err(e)),
+            None => {
+                writer.finish().ok()?;
+                let buffer = std::mem::take(writer.get_mut());
+                Some(Ok(buffer))
+            }
+        });
+        let body_stream = futures::stream::iter(body_iter);
```
Comment (westonpace) [inline rust/lancedb/src/remote/table.rs:75]: Very cool, pity no async writer but I suppose we could wrap in `spawn_blocking` calls at some point in the future.

File: rust/lancedb/src/remote/table.rs [hunk 2]
```diff
@@ -149,82 +173,199 @@ impl<S: HttpSend> TableInternal for RemoteTable<S> {
...
+        })
     }
     async fn plain_query(
         &self,
         _query: &Query,
         _options: QueryExecutionOptions,
     ) -> Result<DatasetRecordBatchStream> {
-        todo!()
+        Err(Error::NotSupported {
+            message: "plain_query is not yet supported on LanceDB cloud.".into(),
```
Comment (westonpace) [inline rust/lancedb/src/remote/table.rs:229]: "not yet" for this one.  You think it is more likely to be added at some point than the others?

File: rust/lancedb/src/remote/table.rs [hunk 3]
```diff
@@ -149,82 +173,199 @@ impl<S: HttpSend> TableInternal for RemoteTable<S> {
...
+            .post(&format!("/table/{}/delete/", self.name))
+            .json(&body);
+        let response = self.client.send(request).await?;
+        self.check_table_response(response).await?;
+        Ok(())
     }
     async fn create_index(&self, _index: IndexBuilder) -> Result<()> {
-        todo!()
+        Err(Error::NotSupported {
+            message: "plain_query is not yet supported on LanceDB cloud.".into(),
```
Comment (westonpace) [inline rust/lancedb/src/remote/table.rs:264]:
```suggestion
            message: "create_index is not yet supported on LanceDB cloud.".into(),
```

File: rust/lancedb/src/remote/table.rs [hunk 4]
```diff
@@ -149,82 +173,199 @@ impl<S: HttpSend> TableInternal for RemoteTable<S> {
...
     async fn list_indices(&self) -> Result<Vec<IndexConfig>> {
-        todo!()
+        Err(Error::NotSupported {
+            message: "list_indices is not yet supported.".into(),
+        })
     }
     async fn table_definition(&self) -> Result<TableDefinition> {
-        todo!()
+        Err(Error::NotSupported {
+            message: "plain_query is not supported on LanceDB cloud.".into(),
```
Comment (westonpace) [inline rust/lancedb/src/remote/table.rs:318]:
```suggestion
            message: "table_definition is not supported on LanceDB cloud.".into(),
```

File: rust/lancedb/src/remote/table.rs [hunk 5]
```diff
@@ -324,4 +480,263 @@ mod tests {
...
+            .unwrap();
+
+        let mut expected_body = Vec::new();
+        {
+            let mut writer =
+                arrow_ipc::writer::StreamWriter::try_new(&mut expected_body, &data_ref.schema())
+                    .unwrap();
+            writer.write(&data_ref).unwrap();
+            writer.finish().unwrap();
+        }
```
Comment (westonpace) [inline rust/lancedb/src/remote/table.rs:545]: Maybe a helper function for this?  record_batch_to_bytes?

File: rust/lancedb/src/remote/table.rs [hunk 6]
```diff
@@ -324,4 +480,263 @@ mod tests {
...
```
Comment (westonpace) [inline rust/lancedb/src/remote/table.rs:603]: record_batch_to_bytes?

File: rust/lancedb/src/remote/table.rs [hunk 7]
```diff
@@ -324,4 +480,263 @@ mod tests {
...
+
+        // Check we serialize data correctly.
+        let mut expected_body = Vec::new();
+        {
+            let mut writer =
+                arrow_ipc::writer::StreamWriter::try_new(&mut expected_body, &batch.schema())
+                    .unwrap();
+            writer.write(&batch).unwrap();
+            writer.finish().unwrap();
+        }
```
Comment (westonpace) [inline rust/lancedb/src/remote/table.rs:719]: record_batch_to_bytes?

**Review by westonpace — APPROVED** [review]
Just some minor nits

File: rust/lancedb/src/remote/table.rs (hunk 2 above)
Comment (wjones127) [reply rust/lancedb/src/remote/table.rs:229]: Well not necessarily. I'm going to fix this up in my follow up PR when I implement query.

**Pull request merged**